### PR TITLE
MM-45869: Global RHS, App Bar (a Threads Everywhere effort)

### PR DESCRIPTION
**Pull request by calebroseland**
CleanShot 2022-08-30 at 15 54 26@2x

(Current Stage: Ready for Review)

#### Summary
- Globalizes RHS and App Bar by hoisting the mounting points to Root, where Global Header and other root components live.   
- Adds a **Grid-based layout** to facilitate simpler RHS handling across products and easier responsive layouts in the future
**This will require an extra-thorough review and testing given foundational aspect of this change**
- Reworks AppBar to better support RHS registrations (see `PluginRegistry.registerAppBarComponent`)
  - Multiple registrations per-plugin now supported!
  **This should also be tested extra-thoroughly**
- Adds a `supportedProductIds` pattern to opt-into showing an RHS registration in certain products (4/5)
- Adds a `showAppBar` alongside `showTeamSidebar` for products to opt-into showing AppBar (5/5)
- Adds actions to `ProductApi` to support interactions with Global RHS (e.g. selecting a thread, closing RHS, selecting if its open/or not, etc.) 

None of this is behind a feature flag because of the foundational nature of these changes and the overwhelming complexity it would require. 

The presence of the various RHS actions in the ProductApi act as an indicator that Global RHS is supported. (Feature-detection paradigm instead of hard-coded version detection).

In order for an AppBar Registration to be shown in AppBar in non-Channels products, a mutual (two-way) opt-in must occur:
- The current product must opt-in to AppBar (`showAppBar: true` when registering the product)
- The RHS registration must define which products it supports (`supportedProductIds: [null, 'playbooks']` in `registerAppBarComponent`)

Certain "core" RHS views will be globally available. e.g. dispatch `SELECT_POST` from anywhere and the existing `RhsThread`-`ThreadViewer` infrastructure will open in RHS. Additional definitions or behaviors will be needed to constrain certain RHS views from crossing into unscoped pages (e.g. Channel Notes RHS should only work in Channels)

All is up for discussion, changes, and refinement.

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-45869

#### Related Pull Requests
- https://github.com/mattermost/mattermost-plugin-calls/pull/176
#### UX/Screenshots
- https://user-images.githubusercontent.com/11724372/196594899-938dee3d-01aa-42ad-b90a-485f8aacb101.mp4

	- RHS-expanded clickaway handling has changed from a fixed-position backdrop div to a true clickaway handler. Clicks from within Global Header, RHS, AppBar, and modals are ignored.
	- RHS-expanded will "snap" its width to take over the center channel area (grid-areas: `rhs` & `center`) at certain viewport widths that would otherwise only leave a small portion of `grid-area: center` visible. The RHS-expanded transition is disabled in this scenario.

#### Release Note

```release-note
UI: new grid-based layout; RHS/AppBar made global/cross-product

Breaking changes:
- Plugins, `PluginRegistry.registerCustomRoute`: when you register a custom route component, you must specify a CSS `grid-area` in order for it to be placed properly into the root layout. (Recommended: `grid-area: center`.)
```

### CODE REVIEW

File: packages/types/src/products.ts [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+/**
+ * - `null` - explicitly Channels
+ * - `string` - uuid - any other product
+ */
+export type ProductIdentifier = null | string;
+
+/** @see {@link ProductIdentifier} */
+export type ProductScope = ProductIdentifier | ProductIdentifier[];
+
+export const inScope = (scope: ProductScope, productId: ProductIdentifier, pluginId?: string) => {
```
Comment (hmhealey) [inline packages/types/src/products.ts:13]: Does this need to live here, or can it be kept in the web app? I'm not sure where we would use this that isn't just in the web app. 0/5 though

File: components/app_bar/app_bar.tsx [hunk 2]
```diff
@@ -1,68 +1,104 @@
...
+        return x;
+    });
+
+    if (!items.some((x) => Boolean(x) && x !== divider)) {
+        return null;
+    }
+
+    return (
+        <div className={'app-bar'}>
+            <div
```
[the remark this answers is not in the document]
Comment (calebroseland) [reply components/app_bar/app_bar.tsx:80]: Preserving the layered transparent background. (Root bg is white.) I tried to minimize it down to one, but multiple background colors aren't easily possible in CSS.

We could hoist various backgrounds, too, but I didn't want to go down that trail just yet.

File: components/app_bar/app_bar.tsx [hunk 3]
```diff
@@ -1,68 +1,104 @@
...
+                css={`
```
[the remark this answers is not in the document]
Comment (calebroseland) [reply components/app_bar/app_bar.tsx:81]: This is "lightweight" styled-components ([css-prop](https://styled-components.com/docs/api#css-prop)).

File: components/app_bar/app_bar_binding.tsx [hunk 4]
```diff
@@ -19,6 +19,10 @@ import {DoAppCallResult} from 'types/apps';
 
 import OverlayTrigger from 'components/overlay_trigger';
 
+export const isAppBinding = (x: unknown): x is AppBinding => {
```
[the remark this answers is not in the document]
Comment (calebroseland) [reply components/app_bar/app_bar_binding.tsx:22]: 2/5 I think it's generally better to reach for `unknown` (https://dev.to/tomdohnal/typescript-any-vs-unknown-a-deep-dive-3iem), but for cases like this where we're casting anyway, I'm not sure it makes any difference (0/5: either way here)

File: components/backstage/backstage_controller.jsx [hunk 5]
```diff
@@ -206,8 +204,15 @@ export default class BackstageController extends React.PureComponent {
...
+                    <GlobalStyleBackstage/>
                 </div>
-            </div>
+            </>
         );
     }
 }
+
+const GlobalStyleBackstage = createGlobalStyle`
+    #root > :not(.backstage-navbar):not(.backstage-body) {
```
[the remark this answers is not in the document]
Comment (calebroseland) [reply components/backstage/backstage_controller.jsx:215]: This was one I was going to come back to, and that's a good point. I'll switch this to to hide by explicit selectors.

File: components/root/root.tsx [hunk 6]
```diff
@@ -442,6 +461,20 @@ export default class Root extends React.PureComponent<Props, State> {
         }
     }
 
+    setRootMeta = () => {
```
[the remark this answers is not in the document]
Comment (calebroseland) [reply components/root/root.tsx:464]: [Container queries](https://caniuse.com/css-container-queries) aren't quite here yet, so the motivation is to better support reacting to whether or not RHS is open/expanded in siblings/parents where necessary.

File: components/root/root.tsx [hunk 7]
```diff
@@ -575,18 +610,29 @@ export default class Root extends React.PureComponent<Props, State> {
...
+                                        let pluggable = (
+                                            <Pluggable
+                                                pluggableName={'Product'}
+                                                subComponentName={'mainComponent'}
+                                                pluggableId={product.id}
+                                                webSocketClient={webSocketClient}
+                                                css={product.wrapped ? undefined : {gridArea: 'center'}}
+                                            />
+                                        );
+                                        if (product.wrapped) {
```
[the remark this answers is not in the document]
Comment (calebroseland) [reply components/root/root.tsx:623]: I was experimenting with this, but could easily remove it—it's to have a flattened `mainComponent` to make use of the `lhs` grid  area if a product so-desired, also to support less nesting and/or potentially better semantics (e.g. a less nested `<main/>`).

File: components/root/root.tsx [hunk 8]
```diff
@@ -597,6 +643,7 @@ export default class Root extends React.PureComponent<Props, State> {
                                         <Pluggable
                                             pluggableName={'CustomRouteComponent'}
                                             pluggableId={plugin.id}
+                                            css={{gridArea: 'center'}}
```
[the remark this answers is not in the document]
Comment (calebroseland) [reply components/root/root.tsx:646]: This is styled-components' css prop being used as a fallback that specifies a `className` that would route the output of this pluggable to `grid-area: center`  IF the registered component has a `className` prop (or uses the `...rest` pattern).

[the remark this answers is not in the document]
Comment (calebroseland) [reply components/root/root.tsx:646]: The `grid-template` is in `_structure.scss`

File: components/sidebar_right/sidebar_right.tsx [hunk 9]
```diff
@@ -120,12 +124,14 @@ export default class SidebarRight extends React.PureComponent<Props, State> {
     componentDidMount() {
         window.addEventListener('resize', this.determineTransition);
         document.addEventListener('keydown', this.handleShortcut);
+        document.addEventListener('mousedown', this.handleClickOutside);
```
[the remark this answers is not in the document]
Comment (calebroseland) [reply components/sidebar_right/sidebar_right.tsx:127]: I agree 👍, but would probably vote to defer that optimization for this PR.

File: components/sidebar_right/sidebar_right.tsx [hunk 10]
```diff
@@ -244,33 +257,36 @@ export default class SidebarRight extends React.PureComponent<Props, State> {
...
+                                    getFocus={this.getSearchBarFocus}
+                                    channelDisplayName={channelDisplayName}
+                                >
+                                    {content}
+                                </Search>
+                            </div>
+                        </div>
+                    </>
+                )}
+                <AppBar/>
```
[the remark this answers is not in the document]
Comment (calebroseland) [reply components/sidebar_right/sidebar_right.tsx:288]: No, it could be a sibling and work just fine. 0/5 on that

`AppBar` has its own `return null` checks. 3/5 I think it's good that those checks are co-located with the main implementation, rather than add conditionals and more logic into root.

File: packages/types/package.json [hunk 11]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "@mattermost/types",
-  "version": "7.1.0",
+  "version": "7.2.0",
```
[the remark this answers is not in the document]
Comment (calebroseland) [reply packages/types/package.json:3]: Yeah, this slipped in when I was doing some testing. Was going to remove.

File: packages/types/src/products.ts [hunk 12]
```diff
@@ -0,0 +1,23 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+/**
+ * - `null` - explicitly Channels
+ * - `string` - uuid - any other product
+ */
+export type ProductIdentifier = null | string;
```
[the remark this answers is not in the document]
Comment (calebroseland) [reply packages/types/src/products.ts:8]: I started down a path of converting this to a meaningful id, but ended up reverting it to minimize changes as this was already an established pattern. 

Though I could see this might be a good time to change it if we were going to. Let's bring it up in web guild perhaps and get some thoughts.

File: plugins/registry.js [hunk 13]
```diff
@@ -778,7 +778,18 @@ export default class PluginRegistry {
...
+        baseURL,
+        switcherIcon,
+        switcherText,
+        switcherLinkURL,
+        mainComponent,
+        headerCentreComponent = () => null,
+        headerRightComponent = () => null,
+        showTeamSidebar = false,
+        showAppBar = false,
+        wrapped = true,
```
[the remark this answers is not in the document]
Comment (calebroseland) [reply plugins/registry.js:791]: 👍

File: plugins/registry.js [hunk 14]
```diff
@@ -778,7 +778,18 @@ export default class PluginRegistry {
     // - showTeamSidebar - A flag to display or hide the team sidebar in products. Defaults to false (hidden).
     // All parameters are required.
     // Returns a unique identifier.
-    registerProduct(baseURL, switcherIcon, switcherText, switcherLinkURL, mainComponent, headerCentreComponent = () => null, headerRightComponent = () => null, showTeamSidebar = false) {
+    registerProduct(
```
[the remark this answers is not in the document]
Comment (calebroseland) [reply plugins/registry.js:781]: I would love to see that, too, though there's a lot of methods with many args. Perhaps in separate effort we could "object-ize" a parallel PluginRegistry and soft deprecate this one.

File: sass/base/_structure.scss [hunk 15]
```diff
@@ -89,13 +85,96 @@ body {
...
+    ;
+
+    > :only-child {
+        grid-area: center;
+        width: 100%;
+    }
+
+
+    .announcement-bar {
+        grid-area: announcement;
```
[the remark this answers is not in the document]
Comment (calebroseland) [reply sass/base/_structure.scss:107]: 1/5 I think here probably makes more sense for the time being. I started experimenting with what felt better, and didn't necessarily like the distributed approach, so I'll try to consolidate a bit.

File: sass/layout/_team-sidebar.scss [hunk 16]
```diff
@@ -166,22 +145,6 @@
         --sidebar-teambar-bg-rgb: var(--sidebar-header-bg-rgb);
         --sidebar-header-text-color-rgb: var(--sidebar-text-rgb);
 
-        // adjust height of team sidebar
-        height: calc(100% - 40px);
         border-right: solid 1px rgba(var(--center-channel-color-rgb), 0.08);
-
-        body.announcement-bar--fixed & {
-            height: calc(100% - #{$announcement-bar-height} - 40px);
```
[the remark this answers is not in the document]
Comment (calebroseland) [reply sass/layout/_team-sidebar.scss:174]: I would expect there to be some edge cases that we'll need to test for/try to find. May be hard to catch every last one, but I'll try to do some searching based on these indicator classes for any similar ones that may need to be removed (plugins, other products, etc.)

File: types/store/plugins.ts [hunk 17]
```diff
@@ -76,6 +77,10 @@ export type Menu = {
 export type PluginComponent = {
     id: string;
     pluginId: string;
+
+    /** @default null - which means 'channels'*/
+    supportedProductIds?: ProductScope;
```
[the remark this answers is not in the document]
Comment (calebroseland) [reply types/store/plugins.ts:82]: The default param in the registry call for this is `null`, so if unspecified, it will follow the current behavior of Channels-only, but if you manually specify `undefined`, `inScope` will return false. Not sure if that level of granularity is useful, but given the current significance of `null`, I thought it was good to keep a "nothing" option.

File: packages/types/src/products.ts (hunk 1 above)
Comment (calebroseland) [reply packages/types/src/products.ts:13]: While I was drafting these, the thought was perhaps usage in plugins/products, but I don't think that's too likely. Slightly in favor of boolean-utils co-located with the types, but also 0/5 so I'll just go ahead and move them back.

File: components/app_bar/app_bar.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (hmhealey) [reply components/app_bar/app_bar.tsx:77]: Oh, do you mean the colour blending we use to get the background colour? I guess we'll have to wait for [`color-mix`](https://developer.mozilla.org/en-US/docs/Web/CSS/color_value/color-mix) to get rid of the extra div

File: components/app_bar/app_bar.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (hmhealey) [reply components/app_bar/app_bar.tsx:78]: Oh I see. Is that not causing an error for you in your editor? I see that we have the babel plugin for it enabled, so it must compile correctly, but VS Code was complaining to me when I tried to do that

[the remark this answers is not in the document]
Comment (calebroseland) [reply components/app_bar/app_bar.tsx:78]: [The type fix is in this PR, too](https://github.com/mattermost/mattermost-webapp/pull/11019/files#diff-3fae5af5ed6bb90a338db7a2c1d4af190366c3a5e70d537ec697c87d859d094d). Sometimes I need to reload/restart VS Code for type changes to take effect.

File: components/root/root.tsx (hunk 8 above)
[the remark this answers is not in the document]
Comment (hmhealey) [reply components/root/root.tsx:647]: Ah, I didn't realize how grid-templates worked when I wrote that.

On a related note, it's weird to me to see the `css` prop from styled-components contain an object. I wonder if it performs better than using `style` due to how it presumably adds a class that applies the style rather than doing it inline

File: components/sidebar_right/sidebar_right.tsx (hunk 10 above)
[the remark this answers is not in the document]
Comment (hmhealey) [reply components/sidebar_right/sidebar_right.tsx]: Yeah, I agree with not adding it to the root. I do want to figure out a good pattern for conditionally mounting objects with as little overhead as possible while the component isn't visible, but that's outside the scope for this. There's a similar issue with menus and modals where we end up calculating a bunch of state for them and just rendering null most of the time since they aren't visible, and it'd be nice to tackle that more broadly

File: plugins/registry.js (hunk 14 above)
[the remark this answers is not in the document]
Comment (hmhealey) [reply plugins/registry.js:785]: Either that or I was thinking of making a helper that converts a potential array of args into an object. That's something I've wanted in the client package as well since some of those API calls are similarly getting massive. Maybe that'd be a good distraction one day since I don't think it'd be super difficult to do

File: sass/layout/_team-sidebar.scss (hunk 16 above)
[the remark this answers is not in the document]
Comment (hmhealey) [reply sass/layout/_team-sidebar.scss:174]: I guess we could even do some ugly hacks like overriding plugin styles here with an `!important` for the time being, but that's hopefully not needed. I guess we should probably wait and see if we actually run into those issues before doing something like that

File: types/store/plugins.ts (hunk 17 above)
[the remark this answers is not in the document]
Comment (hmhealey) [reply types/store/plugins.ts:82]: Makes sense. This definitely isn't the only area that we have `type | undefined | null` as a type. As long as we don't start needing to differentiate unset vs undefined vs null (since that's possible in objects and arrays), this is probably fine

File: components/app_bar/app_bar.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (hmhealey) [reply components/app_bar/app_bar.tsx:78]: Oh, I see. I just tested on master since I didn't see any other changes 😛

File: components/sidebar_right/sidebar_right.tsx [hunk 18]
```diff
@@ -159,9 +165,28 @@ export default class SidebarRight extends React.PureComponent<Props, State> {
             this.props.actions.setRhsExpanded(false);
         }
 
+        // close when changing teams, close when changing productsRHS UX
```
Comment (mickmister) [inline components/sidebar_right/sidebar_right.tsx:168]: Is "RHS UX" meant to be at the end of this comment?

File: sass/base/_structure.scss [hunk 19]
```diff
@@ -89,13 +85,101 @@ body {
...
+    overflow: hidden;
+
+    --columns: min-content min-content minmax(385px, 1fr) minmax(0, auto) min-content;
+    grid-template:
+        'announcement announcement announcement announcement announcement' min-content
+        'header header header header header' min-content
+        'team-sidebar lhs center rhs app-sidebar'
+        'footer footer footer footer footer' min-content
+        / var(--columns);
+    ;
```
Comment (mickmister) [inline sass/base/_structure.scss:103]: I didn't know this was possible. It looks awesome!

File: plugins/registry.ts [hunk 20]
```diff
@@ -62,64 +78,85 @@ const standardizeRoute = (route) => {
...
+        icon,
+        action,
+        dropdownText,
+        tooltipText,
+    }: {
+        icon: ReactResolvable;
+        action: PluginComponent['action'];
+        dropdownText: ReactResolvable;
+        tooltipText: ReactResolvable;
+    }) => {
```
Comment (mickmister) [inline plugins/registry.ts:159]: Amazing work on this file! Since this is now TS, I wonder if we can somehow export `PluginRegistry` through `packages/types` for plugins to import?

File: components/sidebar_right/sidebar_right.tsx (hunk 18 above)
Comment (calebroseland) [reply components/sidebar_right/sidebar_right.tsx]: typo

File: plugins/registry.ts (hunk 20 above)
Comment (calebroseland) [reply plugins/registry.ts:159]: Long term goal, for sure! This is a step towards that end, but still some work to do before that, I believe.

File: components/root/root.tsx (hunk 8 above)
[the remark this answers is not in the document]
Comment (calebroseland) [reply components/root/root.tsx:677]: It can take an object or css string in the same way that `css` can accept either: 

```tsx
import {css} from 'styled-components';
const style: FlattenSimpleInterpolation = css({color: 'red'});
const other: FlattenSimpleInterpolation = css`
    color: red;
`;
```


Under the hood, I believe the css prop is precompiled by babel, so I think the performance is the same as with `styled` wrapping:
```tsx
const thing = styled(Pluggable)`
	color: red;
`;
```

File: components/sidebar_right/sidebar_right.tsx (hunk 10 above)
[the remark this answers is not in the document]
Comment (calebroseland) [reply components/sidebar_right/sidebar_right.tsx]: Yeah there's definitely a balance. (Hoisted rendering conditionals in the parent vs in the child with return null.) Would be nice to have a standard/recommended pattern.

File: plugins/registry.js (hunk 13 above)
[the remark this answers is not in the document]
Comment (calebroseland) [reply plugins/registry.js:981]: added more detail to the [param doc](https://github.com/mattermost/mattermost-webapp/blob/global-apps-bar/types/store/plugins.ts#L209-L215)

File: components/root/root.tsx (hunk 7 above)
[the remark this answers is not in the document]
Comment (hmhealey) [reply components/root/root.tsx:654]: I'd hope that, in the long run, the core app itself could do that management itself in all cases, but I guess we'll see how this is actually used in practice later on.

I definitely appreciate that it's documented in the definition for `ProductWrapper` now though so people will know what to do with it without having to dig through code

File: components/root/root.tsx (hunk 8 above)
[the remark this answers is not in the document]
Comment (hmhealey) [reply components/root/root.tsx:677]: That seems a bit weird to me to see `styled` being used to wrap a React component, especially one that doesn't take a `style` prop. It may also result in a `className` prop being added since I don't remember which styled-components uses, but that confusion applies either way.

Not blocking, but does this require that the component provided by the plugin accepts a style prop? It seems like this might not work if the provided component doesn't because the styles wouldn't be applied to anything.

File: packages/types/src/products.ts (hunk 1 above)
Comment (hmhealey) [reply packages/types/src/products.ts]: Sounds good to me. We really need to get around to figuring out where those sort of shared utilities go one of these days...

**Review by hmhealey — APPROVED** [review]
A few followup comments, but nothing that specifically requires changes, so this looks good to me

File: components/global_header/left_controls/product_menu/product_branding/product_branding.tsx [hunk 21]
```diff
@@ -19,7 +19,7 @@ const ProductBrandingContainer = styled.div`
 
 const ProductBranding = (): JSX.Element => {
     const products = useProducts();
-    const currentProductID = useCurrentProductId(products);
+    const currentProductID = useCurrentProductId();
     const currentProduct = products?.find((product) => product.id === currentProductID);
```
Comment (koox00) [inline components/global_header/left_controls/product_menu/product_branding/product_branding.tsx:23]: would it make sense to just use `const currentProduct = useCurrentProduct()`;

File: components/rhs_thread/rhs_thread.tsx [hunk 22]
```diff
@@ -25,6 +27,15 @@ const RhsThread = ({
     channel,
     previousRhsState,
 }: Props) => {
+    const dispatch = useDispatch();
+    const currentTeam = useSelector(getCurrentTeam);
```
Comment (koox00) [inline components/rhs_thread/rhs_thread.tsx:31]: do we need this? since we add `currentTeam` in `mapStateToProps`

File: utils/products.ts [hunk 23]
```diff
@@ -1,19 +1,53 @@
...
+import {useSelector} from 'react-redux';
 
-    for (let i = 0; i < products.length; i++) {
-        const product = products[i];
-        if (pathname.startsWith(product.baseURL)) {
-            return product.id;
-        }
-    }
+import {ProductComponent} from 'types/store/plugins';
+import {selectProducts, selectCurrentProductId, selectCurrentProduct} from 'selectors/products';
```
Comment (koox00) [inline utils/products.ts:9]: Are we creating a circular dependency here since `selectCurrentProduct{Id}` depend on `getCurrentProduct`?

BTW, do we need `getCurrentProduct{Id}`?

File: plugins/registry.ts [hunk 24]
```diff
@@ -826,65 +1029,102 @@ export default class PluginRegistry {
...
             name: 'AppBar',
-            data,
+            data: {
+                id,
+                pluginId: this.id,
+                iconUrl,
+                tooltipText: resolveReactElement(tooltipText),
+                supportedProductIds,
+                ...registeredRhsComponent ? {
+                    action: () => store.dispatch(registeredRhsComponent.toggleRHSPlugin),
```
[the remark this answers is not in the document]
Comment (calebroseland) [reply plugins/registry.ts:1111]: Indeed this is not ideal, but `store` should be reference-safe in the same way that `stores/redux_store` is. (It's also the same pattern that Playbooks [currently uses](https://github.com/mattermost/mattermost-plugin-playbooks/blob/master/webapp/src/index.tsx#L196-L199).) 

This is necessary because `action` somewhat deceivingly named—it's just a callback that receives a channel and channelMember (https://github.com/mattermost/mattermost-webapp/blob/global-apps-bar/components/app_bar/app_bar_plugin_component.tsx#L103).

File: components/root/root.tsx (hunk 8 above)
[the remark this answers is not in the document]
Comment (calebroseland) [reply components/root/root.tsx:680]: `styled(...)`-wrapping a component, or using the `css` prop, will yield a `className` addition. Indeed, `className` or rest attrs would need to be supported in the registered `CustomRouteComponent`. (This is supported via `Pluggable`'s rest-spread props.)

File: components/rhs_thread/rhs_thread.tsx (hunk 22 above)
Comment (calebroseland) [reply components/rhs_thread/rhs_thread.tsx]: nope, thanks for noticing that dupe/unnecessary selector!

File: utils/products.ts (hunk 23 above)
Comment (calebroseland) [reply utils/products.ts:9]: Indeed there are a lot of cross-references between these two modules, but no the individual methods shouldn't create any circular dependencies. Flow: `use` -> `select` -> `get`.

When I refactored these, I made sure there were `get`, `select`, and `use` varieties for both `ProductComponent` and `ProductIdentifier`. These are used in a lot of different contexts (connected, functional components, etc.) and I felt the flexibility was worth supporting even though it's not currently used.

That said, I'm 0/5 on leaving `getCurrentProductId` in or not—doesn't matter to me.

File: packages/types/src/products.ts (hunk 1 above)
Comment (calebroseland) [reply packages/types/src/products.ts]: ^ 💯

**Review by koox00 — APPROVED** [review]

**Comment by calebroseland** [comment]
Will address a few related E2E failures https://mattermost-cypress-report.s3.amazonaws.com/252775-6c26a6c-onprem-ent-webapp-pr-11019-mm-ee-test:6c26a6c/mochawesome.html

**Comment by calebroseland** [comment]
@jgilliam17 everything should be good to go for final QA.

**Review by jgilliam17 — APPROVED** [review]
Thanks @calebroseland 
I reviewed [E2E report](https://automation-dashboard.vercel.app/cycles/cba427e9-7f5b-432b-977e-a52abae96623), one possible failure related to changes, not reproducing manually `MM-248_2` 
Manually tested rhs for boards and playbooks - no issues. Also checked pinned, saved, search mentions etc. working as before.
@DHaussermann or @Flinch Can you please test the plugins portion, breaking changes are listed in the notes 
Breaking changes:
- Plugins, `PluginRegistry.registerCustomRoute`: when you register a custom route component, you must specify a CSS `grid-area` in order for it to be placed properly into the root layout. (Recommended: `grid-area: center`.)

**Comment by lieut-data** [comment]
As discussed and agreed upon at https://community.mattermost.com/core/pl/4gyeua1xr3bu8rzz5s3oicxiha, merging to get this on community to maximize immediate coverage.

**Comment by DHaussermann** [comment]
@calebroseland I don't believe this custom path functionality is used by any of the plugins integrations team maintains. Is it possible for you to ad something that validates this with functional testing to the demo plugin?  Otherwise do you know of another way to validate this change?
